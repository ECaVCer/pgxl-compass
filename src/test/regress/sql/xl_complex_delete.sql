-- Some complex delete from/DELETE queries also known to have problems
-- Check delete from with inherited target and an inherited source table
create table xl_d_foo(f1 int, f2 int);
create table xl_d_foo2(f3 int) inherits (xl_d_foo);
create table xl_d_bar(f1 int, f2 int);
create table xl_d_bar2(f3 int) inherits (xl_d_bar);
create table xl_d_foo1(f1 int, f2 int);--non inherited or inheriting table

insert into xl_d_foo values(1,1);
insert into xl_d_foo values(3,3);
insert into xl_d_foo1 values(1,1);
insert into xl_d_foo1 values(3,3);
insert into xl_d_foo2 values(2,2,2);
insert into xl_d_foo2 values(3,3,3);
insert into xl_d_bar values(1,1);
insert into xl_d_bar values(2,2);
insert into xl_d_bar values(3,3);
insert into xl_d_bar values(4,4);
insert into xl_d_bar2 values(1,1,1);
insert into xl_d_bar2 values(2,2,2);
insert into xl_d_bar2 values(3,3,3);
insert into xl_d_bar2 values(4,4,4);

create function xl_cd_nodename_from_id(integer) returns name as $$
declare 
	n name;
BEGIN
	select node_name into n from pgxc_node where node_id = $1;
	RETURN n;
END;$$ language plpgsql;

select xl_cd_nodename_from_id(xc_node_id), * from xl_d_foo; 

select xl_cd_nodename_from_id(xc_node_id), * from only xl_d_foo; 

select xl_cd_nodename_from_id(xc_node_id), * from xl_d_foo2;

select xl_cd_nodename_from_id(xc_node_id), * from xl_d_bar;

select xl_cd_nodename_from_id(xc_node_id), * from only xl_d_bar;

select xl_cd_nodename_from_id(xc_node_id), * from xl_d_bar2;

delete from xl_d_bar where f1 in (select f1 from xl_d_foo);-- fail as complex query part has parent table

delete from xl_d_bar where f1 in (select f1 from only xl_d_foo);-- pass as complex query part has only parent table

select xl_cd_nodename_from_id(xc_node_id), * from xl_d_bar;

select xl_cd_nodename_from_id(xc_node_id), * from only xl_d_bar;

delete from only xl_d_bar where f1 in (select f1 from xl_d_foo);-- fail as complex query part has parent table

delete from only xl_d_bar where f1 in (select f1 from only xl_d_foo);-- pass as complex query part has only parent table

select xl_cd_nodename_from_id(xc_node_id), * from xl_d_bar;

select xl_cd_nodename_from_id(xc_node_id), * from only xl_d_bar;

-- checking independent working of union all
select f1 from only xl_d_foo;

select f1+3 from only xl_d_foo;

select f1 from only xl_d_foo union all select f1+3 from only xl_d_foo;

-- Check delete from with inherited target and an appendrel subquery
-- All below cases fail as union all in complex delete from subquery is not yet supported.
delete from xl_d_bar
where f1 in ( select f1 from xl_d_foo union all select f1+3 from xl_d_foo );--fail

delete from xl_d_bar 
where f1 in ( select f1 from only xl_d_foo union all select f1+3 from xl_d_foo );--fail

delete from xl_d_bar 
where f1 in ( select f1 from only xl_d_foo union all select f1+3 from only xl_d_foo );--pass

delete from only xl_d_bar 
where f1 in ( select f1 from only xl_d_foo union all select f1+3 from only xl_d_foo );--pass

-- xl_d_foo1 is a non-inheriting or inherited table, still union in subquery fails
delete from xl_d_bar
where f1 in ( select f1 from xl_d_foo1 union all select f1+3 from xl_d_foo1 );--pass

delete from only xl_d_bar
where f1 in ( select f1 from xl_d_foo1 union all select f1+3 from xl_d_foo1 );--pass

drop function xl_cd_nodename_from_id(integer);
drop table xl_d_foo2;
drop table xl_d_bar2;
drop table xl_d_bar;
drop table xl_d_foo;
drop table xl_d_foo1;
