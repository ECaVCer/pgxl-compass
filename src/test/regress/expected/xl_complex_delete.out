-- Some complex delete from/DELETE queries also known to have problems
-- Check delete from with inherited target and an inherited source table
create table xl_d_foo(f1 int, f2 int);
create table xl_d_foo2(f3 int) inherits (xl_d_foo);
create table xl_d_bar(f1 int, f2 int);
create table xl_d_bar2(f3 int) inherits (xl_d_bar);
create table xl_d_foo1(f1 int, f2 int);--non inherited or inheriting table
insert into xl_d_foo values(1,1);
insert into xl_d_foo values(3,3);
insert into xl_d_foo1 values(1,1);
insert into xl_d_foo1 values(3,3);
insert into xl_d_foo2 values(2,2,2);
insert into xl_d_foo2 values(3,3,3);
insert into xl_d_bar values(1,1);
insert into xl_d_bar values(2,2);
insert into xl_d_bar values(3,3);
insert into xl_d_bar values(4,4);
insert into xl_d_bar2 values(1,1,1);
insert into xl_d_bar2 values(2,2,2);
insert into xl_d_bar2 values(3,3,3);
insert into xl_d_bar2 values(4,4,4);
create function xl_cd_nodename_from_id(integer) returns name as $$
declare 
	n name;
BEGIN
	select node_name into n from pgxc_node where node_id = $1;
	RETURN n;
END;$$ language plpgsql;
select xl_cd_nodename_from_id(xc_node_id), * from xl_d_foo; 
 xl_cd_nodename_from_id | f1 | f2 
------------------------+----+----
 datanode_1             |  1 |  1
 datanode_2             |  3 |  3
 datanode_1             |  2 |  2
 datanode_2             |  3 |  3
(4 rows)

select xl_cd_nodename_from_id(xc_node_id), * from only xl_d_foo; 
 xl_cd_nodename_from_id | f1 | f2 
------------------------+----+----
 datanode_1             |  1 |  1
 datanode_2             |  3 |  3
(2 rows)

select xl_cd_nodename_from_id(xc_node_id), * from xl_d_foo2;
 xl_cd_nodename_from_id | f1 | f2 | f3 
------------------------+----+----+----
 datanode_1             |  2 |  2 |  2
 datanode_2             |  3 |  3 |  3
(2 rows)

select xl_cd_nodename_from_id(xc_node_id), * from xl_d_bar;
 xl_cd_nodename_from_id | f1 | f2 
------------------------+----+----
 datanode_1             |  1 |  1
 datanode_1             |  2 |  2
 datanode_2             |  3 |  3
 datanode_2             |  4 |  4
 datanode_1             |  1 |  1
 datanode_1             |  2 |  2
 datanode_2             |  3 |  3
 datanode_2             |  4 |  4
(8 rows)

select xl_cd_nodename_from_id(xc_node_id), * from only xl_d_bar;
 xl_cd_nodename_from_id | f1 | f2 
------------------------+----+----
 datanode_1             |  1 |  1
 datanode_1             |  2 |  2
 datanode_2             |  3 |  3
 datanode_2             |  4 |  4
(4 rows)

select xl_cd_nodename_from_id(xc_node_id), * from xl_d_bar2;
 xl_cd_nodename_from_id | f1 | f2 | f3 
------------------------+----+----+----
 datanode_1             |  1 |  1 |  1
 datanode_1             |  2 |  2 |  2
 datanode_2             |  3 |  3 |  3
 datanode_2             |  4 |  4 |  4
(4 rows)

delete from xl_d_bar where f1 in (select f1 from xl_d_foo);-- fail as complex query part has parent table
ERROR:  could not plan this distributed delete
DETAIL:  correlated or complex DELETE is currently not supported in Postgres-XL.
delete from xl_d_bar where f1 in (select f1 from only xl_d_foo);-- pass as complex query part has only parent table
select xl_cd_nodename_from_id(xc_node_id), * from xl_d_bar;
 xl_cd_nodename_from_id | f1 | f2 
------------------------+----+----
 datanode_1             |  2 |  2
 datanode_2             |  4 |  4
 datanode_1             |  2 |  2
 datanode_2             |  4 |  4
(4 rows)

select xl_cd_nodename_from_id(xc_node_id), * from only xl_d_bar;
 xl_cd_nodename_from_id | f1 | f2 
------------------------+----+----
 datanode_1             |  2 |  2
 datanode_2             |  4 |  4
(2 rows)

delete from only xl_d_bar where f1 in (select f1 from xl_d_foo);-- fail as complex query part has parent table
ERROR:  could not plan this distributed delete
DETAIL:  correlated or complex DELETE is currently not supported in Postgres-XL.
delete from only xl_d_bar where f1 in (select f1 from only xl_d_foo);-- pass as complex query part has only parent table
select xl_cd_nodename_from_id(xc_node_id), * from xl_d_bar;
 xl_cd_nodename_from_id | f1 | f2 
------------------------+----+----
 datanode_1             |  2 |  2
 datanode_2             |  4 |  4
 datanode_1             |  2 |  2
 datanode_2             |  4 |  4
(4 rows)

select xl_cd_nodename_from_id(xc_node_id), * from only xl_d_bar;
 xl_cd_nodename_from_id | f1 | f2 
------------------------+----+----
 datanode_1             |  2 |  2
 datanode_2             |  4 |  4
(2 rows)

-- checking independent working of union all
select f1 from only xl_d_foo;
 f1 
----
  1
  3
(2 rows)

select f1+3 from only xl_d_foo;
 ?column? 
----------
        4
        6
(2 rows)

select f1 from only xl_d_foo union all select f1+3 from only xl_d_foo;
 f1 
----
  1
  3
  4
  6
(4 rows)

-- Check delete from with inherited target and an appendrel subquery
-- All below cases fail as union all in complex delete from subquery is not yet supported.
delete from xl_d_bar
where f1 in ( select f1 from xl_d_foo union all select f1+3 from xl_d_foo );--fail
ERROR:  could not plan this distributed delete
DETAIL:  correlated or complex DELETE is currently not supported in Postgres-XL.
delete from xl_d_bar 
where f1 in ( select f1 from only xl_d_foo union all select f1+3 from xl_d_foo );--fail
ERROR:  could not plan this distributed delete
DETAIL:  correlated or complex DELETE is currently not supported in Postgres-XL.
delete from xl_d_bar 
where f1 in ( select f1 from only xl_d_foo union all select f1+3 from only xl_d_foo );--pass
delete from only xl_d_bar 
where f1 in ( select f1 from only xl_d_foo union all select f1+3 from only xl_d_foo );--pass
-- xl_d_foo1 is a non-inheriting or inherited table, still union in subquery fails
delete from xl_d_bar
where f1 in ( select f1 from xl_d_foo1 union all select f1+3 from xl_d_foo1 );--pass
delete from only xl_d_bar
where f1 in ( select f1 from xl_d_foo1 union all select f1+3 from xl_d_foo1 );--pass
drop function xl_cd_nodename_from_id(integer);
drop table xl_d_foo2;
drop table xl_d_bar2;
drop table xl_d_bar;
drop table xl_d_foo;
drop table xl_d_foo1;
